### DISCUSSION TITLE
Generic Props and type alias Props

### DISCUSSION THREAD

**Discussion by tsar-boomba** [withastro/roadmap#398] [Proposal]
I am currently trying to make polymorphic components in Astro using the HTMLAttributes type from `astro/types`. I have run into the problem that If you make the interface Props generic, its props type will be `any` when you use the component. Along with this, I would need to be able to use type aliases for props types to allow "extending" other generic types, which interfaces do not allow.

# Example syntax

## Interface Generics
```astro
---
import { HTMLTag } from 'astro/types';

// currently, adding a generic to an interface will cause the props to be `any`
interface Props<Tag extends HTMLTag> {
    component?: Tag
}

const { component = 'div', ...props } = Astro.props
const Element = component;
---
<Element {...Astro.props}>
    <slot />
</Element>
```

## Type Alias For Props

```astro
---
import { HTMLTag, HTMLAttributes } from 'astro/types';

// This allows for the attributes available to match the tag used by the user
// This is not possible with interfaces at all, they cannot extend a generic type
type Props<Tag extends HTMLTag> = {
    component?: Tag
} & HTMLAttributes<Tag>;

const { component = 'div', ...props } = Astro.props
const Element = component;
---
<Element {...Astro.props}>
    <slot />
</Element>
```

**Comment by florian-lefebvre** [comment]
This is now possible: https://docs.astro.build/en/guides/typescript/#polymorphic-type
